### PR TITLE
Locked asset refactor

### PR DESCRIPTION
**Pull request by catalinnnn**

### CODE REVIEW

File: sc-locked-asset-factory/src/lib.rs [hunk 1]
```diff
@@ -118,11 +131,16 @@ pub trait LockedAssetFactory:
...
+                current_block_epoch,
+                &unlock_schedule.unlock_milestones,
+            );
+            let new_unlock_schedule = UnlockSchedule {
+                unlock_milestones: new_unlock_milestones,
+            };
             let _ = self.produce_tokens_and_send(
                 &locked_remaining,
-                &new_unlock_milestones,
+                &new_unlock_schedule,
```
Comment (sasurobert) [inline sc-locked-asset-factory/src/lib.rs:143]: How is this computed : I have 100 Locked MEX, I unlock after first period end and will receive 10MEX and 90 locked mex with release schedule of 15 percent for the next 6 months. After one month I unlock_assets again. 15% of 90 = 13.5 MEX and I should have received 15. In the end if I unlock at every month I will get much less tokens.

File: sc-locked-asset-factory/src/lib.rs [hunk 2]
```diff
@@ -106,10 +116,13 @@ pub trait LockedAssetFactory:
         let locked_token_id = self.locked_asset_token_id().get();
         require!(token_id == locked_token_id, "Bad payment token");
 
-        let attributes = self.get_attributes(&token_id, token_nonce)?;
+        let unlock_schedule = self.get_unlock_schedule_for_sft_nonce(token_nonce).unwrap();
         let current_block_epoch = self.blockchain().get_block_epoch();
-        let unlock_amount =
-            self.get_unlock_amount(&amount, current_block_epoch, &attributes.unlock_milestones);
+        let unlock_amount = self.get_unlock_amount(
```
Comment (sasurobert) [inline sc-locked-asset-factory/src/lib.rs:121]: I think sft.start_epoch should be sent for get_unlock_amount

File: sc-locked-asset-factory/src/lib.rs (hunk 1 above)
Comment (catalinnnn) [reply sc-locked-asset-factory/src/lib.rs:143]: The rest of percents are recalculated when unlocking. A user won't get 15% from 90. He will get the equivalent of 15% from the initial value.

File: sc-locked-asset-factory/src/lib.rs (hunk 2 above)
Comment (catalinnnn) [reply sc-locked-asset-factory/src/lib.rs:121]: Fixed

File: sc-locked-asset-factory/src/lib.rs [hunk 3]
```diff
@@ -91,7 +97,11 @@ pub trait LockedAssetFactory:
         require!(amount > 0, "Zero input amount");
         require!(!schedule.is_empty(), "Empty param");
 
-        let _ = self.produce_tokens_and_send(&amount, &schedule.0, &address, &OptionalArg::None);
+        let unlock_schedule = UnlockSchedule {
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply sc-locked-asset-factory/src/lib.rs:100]: Fixed

File: sc-locked-asset-factory/src/lib.rs (hunk 2 above)
Comment (catalinnnn) [reply sc-locked-asset-factory/src/lib.rs:121]: why? need to calculate the unlock with regard to the current epoch (or at least current month epoch)

File: sc-locked-asset-factory/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (catalinnnn) [reply sc-locked-asset-factory/src/lib.rs:100]: produce_tokens_and_send is still used in two other functions

File: sc-locked-asset-factory/src/lib.rs (hunk 2 above)
Comment (sasurobert) [reply sc-locked-asset-factory/src/lib.rs:97]: yeah. not more needed.

**Review by sasurobert — APPROVED** [review]

**Pull request merged**